packages/core/src/cli/validate/upgradeability-assessment.ts

```diff
@@ -0,0 +1,138 @@
+import { SourceContract } from './validations';
+
+export class ReferenceContractNotFound extends Error {
+  /**
+   * The contract reference that could not be found.
+   */
+  readonly reference: string;
+
+  /**
+   * The fully qualified name of the contract that referenced the missing contract.
+   */
+  readonly origin: string;
+
+  constructor(reference: string, origin: string) {
+    super(`Could not find contract ${reference} referenced in ${origin}.`);
+    this.reference = reference;
+    this.origin = origin;
+    this.name = 'ReferenceContractNotFound';
+  }
+}
+
+interface AnnotationAssessment {
+  upgradeable: boolean;
+  referenceName?: string;
+}
+
+export interface UpgradeabilityAssessment {
+  upgradeable: boolean;
+  referenceContract?: SourceContract;
+  uups?: boolean;
+}
+
+export function getUpgradeabilityAssessment(
+  contract: SourceContract,
+  allContracts: SourceContract[],
+): UpgradeabilityAssessment {
+  const fullContractName = contract.fullyQualifiedName;
+  const contractValidation = contract.validationData[fullContractName];
+
+  const isUUPS = inferUUPS(contract.validationData, fullContractName);
+
+  const annotationAssessment = getAnnotationAssessment(contract);
+  if (annotationAssessment.upgradeable) {
+    let referenceContract = undefined;
+    let isReferenceUUPS = false;
+    if (annotationAssessment.referenceName !== undefined) {
+      referenceContract = getReferenceContract(annotationAssessment.referenceName, contract, allContracts);
+      isReferenceUUPS = inferUUPS(referenceContract.validationData, referenceContract.fullyQualifiedName);
+    }
+
+    return {
+      upgradeable: true,
+      referenceContract: referenceContract,
+      uups: isReferenceUUPS || isUUPS,
+    };
+  } else {
+    const initializable = inferInitializable(contractValidation);
+    return {
+      upgradeable: initializable || isUUPS,
+      uups: isUUPS,
+    };
+  }
+}
+
+function getReferenceContract(reference: string, origin: SourceContract, allContracts: SourceContract[]) {
+  const referenceContracts = allContracts.filter(c => c.fullyQualifiedName === reference || c.name === reference);
+
+  if (referenceContracts.length > 1) {
+    throw new ValidateCommandError(
+      `Found multiple contracts with name ${reference} referenced in ${origin.fullyQualifiedName}.`,
+      () =>
+        `This may be caused by old copies of build info files. Clean and recompile your project, then run the command again with the updated files.`,
+    );
+  } else if (referenceContracts.length === 1) {
+    return referenceContracts[0];
+  } else {
+    throw new ReferenceContractNotFound(reference, origin.fullyQualifiedName);
+  }
+}
+
+function getAnnotationAssessment(contract: SourceContract): AnnotationAssessment {
+  const node = contract.node;
+
+  if ('documentation' in node) {
+    const doc = typeof node.documentation === 'string' ? node.documentation : node.documentation?.text ?? '';
```

Comment:
This line comes from `src/validate/run.ts`. Let's extract it to a shared function used in both locations.